### ISSUE TITLE
Support federated login for guest users (OpenID)

### ISSUE THREAD

**Issue by stevenroose** [go-gitea/gitea#185]
_From @stevenroose on May 26, 2015 9:5_

It bothers many that they have to enter a full form of personal information in order to just file a bug or make a feature request.
OpenID support can fix this and allows guest users to login with they GitHub of Google account f.e.

Ideally, it would also be possible to restrict these "guest"/external users to only be able to make create issues and pull requests and not have their own repositories.

_Copied from original issue: gogits/gogs#1297_

**Comment by stevenroose** [comment]
_From @baby-gnu on February 3, 2016 14:28_

+1

**Comment by stevenroose** [comment]
_From @strk on April 3, 2016 17:9_

+1

**Comment by stevenroose** [comment]
_From @strk on April 5, 2016 14:16_

I'd love to be able to login with my own openid provider (not any of the known commercial providers)

**Comment by stevenroose** [comment]
_From @strk on April 5, 2016 14:35_

May be useful: https://github.com/emanoelxavier/openid2go

**Comment by stevenroose** [comment]
_From @strk on April 6, 2016 12:11_

About how recent changes in OpenID specification are making the "federated" part harder:
https://trac.simpleid.koinic.net/ticket/156#comment:1

I guess "OpenID 2.0" support would be a good start here, and approaching "OpenID Connect" only when the whole "self-descovery" parts are also available.

**Comment by stevenroose** [comment]
Any updates on this?

**Comment by stevenroose** [comment]
_From @strk on May 25, 2016 10:32_

I could start working on this if my pull requests weren't all stuck in the queue :(
The single-man management might be showing its limits here. With @Unknwon being busy things don't seem to move easily.

**Comment by stevenroose** [comment]
_From @strk on May 25, 2016 10:33_

@stevenroose but maybe you could put the "usage restriction" part into a separate issue as I think it's independent of the OpenID support (ie: I'd like to have that also for LDAP-only authentication setup)

**Comment by stevenroose** [comment]
@strk done in https://github.com/gogits/gogs/issues/3130.

And I agree. @Unknwon's single-man management was the reason the Gitea fork came in existence some time ago, until Unknwon appeared again.

I would urge Unknwon to try to select at least another contributor to have permission to review PR's so that the project would move along when he has no time to do all the reviews.

**Comment by stevenroose** [comment]
_From @sbrl on June 29, 2016 10:37_

Has any progress been made here? I've been waiting for this for a while now. I'm using Go Git Service for a personal git server, and I don't have the resources to run anything else that I've seen around so far. I really don't want to end up using unsupported software.

**Comment by stevenroose** [comment]
_From @strk on June 29, 2016 11:35_

I hadn't started, nor I plan to start any soon (and my other PRs are still pending).
@sbrl did you consider giving it a try yourself ? You may go get github.com/yohcop/openid-go/

**Comment by stevenroose** [comment]
_From @sbrl on June 29, 2016 12:24_

@strk I might actually when I get some time. I'd have to read into go first though, which shouldn't be a huge deal. It may not be for some considerable time though, as I'm currently focusing on something for University. I'd be happy for someone else to give it  go in the mean time.

I am worried though that if I gave it a go, my code wouldn't work with any of the other (currently 40) PRs that have been submitted.

**Comment by stevenroose** [comment]
@sbrl That's why you can rebase your code before it is merged ;)

**Comment by stevenroose** [comment]
_From @sbrl on June 29, 2016 15:33_

@stevenroose Eh?

**Comment by stevenroose** [comment]
_From @roidelapluie on August 17, 2016 11:45_

Hello,

OpenID is deprecated since 2014.

We should use OpenID Connect instead. There is the go-openidc library for this.

**Comment by stevenroose** [comment]
_From @strk on August 17, 2016 13:14_

On August 17, 2016 1:45:35 PM CEST, Julien Pivotto <notifications@github.com> wrote:
>Hello,
>
>OpenID is deprecated since 2014.
>
>We should use OpenID Connect instead. There is the go-openidc library
>for this.

Openid Connect with the self-discovery extension would be a full replacement of Openid, but very few server implementations exist. For this reason I think it would still be good to also accept OpenID.
--strk;

Sent from hand-held device, sorry for brevity

**Comment by stevenroose** [comment]
I think OpenID Connect is mostly considered the "latest version" of OpenID.
So yeah, I agree that we should support OpenID in the form of OpenID
Connect.

On Aug 17, 2016 13:45, "Julien Pivotto" <notifications@github.com> wrote:

Hello,

OpenID is deprecated since 2014.

We should use OpenID Connect instead. There is the go-openidc library for
this.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/gogits/gogs/issues/1297#issuecomment-240387969>, or mute
the thread

**Comment by stevenroose** [comment]
_From @roidelapluie on August 18, 2016 13:5_

@strk  Who comes first.... Client implementations or server implementations... OpenID connect is different from OpenID 1.0/2.0. Should this be another ticket?

**Comment by stevenroose** [comment]
_From @strk on August 18, 2016 15:39_

To avoid a race I'd suggest to support both
OpenID 2.0 and OpenID Connect (with the self-discovery extension).

The important thing is to allow people to login using their
own authentication means, basically. That is, identifying users
by a web URL they provide, and doing it in a way that really
allows anyone to setup the self-authentication infrastructure
in an easy way.

**Comment by stevenroose** [comment]
_From @roidelapluie on August 18, 2016 15:44_

My usecase for OpenID Connect is without the self-discovery extension, inside one company's infrastructure, I am not interested by the self discovery. So that is 3 issues?

**Comment by stevenroose** [comment]
_From @strk on August 18, 2016 15:46_

If you want. This ticket title is about "federated login for guest
users" so different usecase from yours.

**Comment by stevenroose** [comment]
_From @bkcsoft on August 18, 2016 16:7_

@strk @roidelapluie It's the same issue, since it also uses OpenID. (Please don't create multiple issue for the same thing :laughing:)

**Comment by stevenroose** [comment]
_From @strk on September 3, 2016 20:20_

If someone wants to help, I've started writing some code in #3595

**Comment by sbrl** [comment]
Waaaaat? I don't understand what's going on here. What's `gitea`? Why has it been forked? Is `gogs` still a thing?

To answer [this comment], I don't think I can since I don't have go installed, I don't know go, I don't understand the go package system, and I don't understand openid :confused: :disappointed:

???

**Comment by tboerger** [comment]
@stevenroose you shouldn't copy everything, a single reference to the issue should be fine.

**Comment by strk** [comment]
@sbrl `gitea` is a fork of `gogs` made to avoid development being blocked by a single person in charge of accepting contributions. Management of `gitea` is in the hands of 3 owners elected yearly by a growing number of maintainers (currently 10). See the CONTRIBUTING.md file for further details.

Binary packages, for those who don't want to build from sources, are in the works, see #31

**Comment by sbrl** [comment]
Ah right! I might switch when a full release is made then.

**Comment by bkcsoft** [comment]
@tboerger One can always clean it up 😉 (Like I just did)

**Comment by tscs37** [comment]
I'm in favor of adding basic OpenID support and like to also propose OpenID Connect support.

That would allow arbitrary federation from any provider that supports OpenID Connect.

**Comment by strk** [comment]
@tscs37 are you up to try at moving some steps toward the implementation of one or the other ?
I've moved the first step for OpenID but it's far from ready (mostly needs UX design), see #271 

As per `OpenID Connect`, do you know of any consumer implementation that *does* support arbitrary provider support ? I'm asking because I don't and would like to spread the word about its existance.

**Comment by stevenroose** [comment]
@strk StackExchange has it (prob not open source). I thought the OSQA had it too, that's an old open-source clone of SE. Or is that basic OpenID?

**Comment by strk** [comment]
I can log into StackExchange with my OpenID-2.0 provider
(not OAuth based). I dont' have an OpenID-Connect provider
to test with StackExchange, but I suspect only hard-coded
providers would be supported there.

**Comment by stevenroose** [comment]
Just to copy from @strk 's PR:

The big advantage I would see from having OpenID is as a way of users to login without having been registered. Just like systems like Discuss allow anyone with an OpenID (even though Discuss requires this to be from a small set of providers) to make contributions in the form of issues or (federated) pull requests.

A possible setup that would benefit greatly from OpenID is the case of single-user usage. Gitea as a self-hosted Git homepage in which only one user has commit access, while he can allow anyone to open issues or request pulls with their OpenID (and a CAPTCHA ofc).

**Comment by strk** [comment]
Ouch, I already commented in the PR, can we continue there ?

**Comment by stevenroose** [comment]
Does #816 allow guest users in some way?

**Comment by strk** [comment]
@stevenroose I guess you meant #618 (OpenID). It depends on what you mean by "guest users".
Logging in via OpenID still needs you to register as a local user, and confirm your email if the service is configured like that.

The registration step is needed to have a unique identifier on the specific node (rather than using the full OpenID URI as an identifier). Also the code to deal with permissions and teams invitations and such are all based on a local "UserID", so registration is also needed for that (to assign an UserID to an OpenID account).

Doing the registration step also lets you associate multiple OpenID URIs to a single account.

BTW, @stevenroose as you mentioned stackexchange, it also supports OpenID-2.0 so you do can login to Gitea now with "openid.stackexchange.com" as your OpenID URI :)

**Comment by sbrl** [comment]
@strk Awesome! So can I restrict auto-registered account to not be able to create repositories too?

**Comment by strk** [comment]
To [my knowledge](https://docs.gitea.io/en-us/config-cheat-sheet/) there's currently no support to specify default permissions for new users, but it'll be great to add that feature !

**Comment by stevenroose** [comment]
@strk @sbrl, yeah that's what I meant with "guest users". Those that only can make issues and submit federated PRs. I.e. not being able to take resources, so that you don't have to trust them not to DOS down your Gitea..

**Comment by strk** [comment]
"taking resources" should be clearly defined.
Opening a PR at the *bare* minimum probably implies opening
a "ticket" record (issue/PR) and probably also the ability
to discuss/react to comments (so more records, issue/pr comments).

I suspect that alone would currently be enough to DOS a Gitea instance.

Anyway I do see also another problem with limiting the space taken
by a user. A fork might probably not take much space until it diverges
too much from the original, but how much is too much ?

I think this discussion (setting limits for users) should be discussed
in a separate ticket, IMHO.

Beside, if I'm not mistaken someone started work on a more fine-grained
set of permissions (check out open issues/PRs)

**Comment by sbrl** [comment]
I'd allow the admin to set the amount of space allowed on a per-group basis or something. My main concern is losing all my server's (rather limited) hard drive space.

And can I migrate to gitea from gogs version 0.10.1.0228? I missed the first release of gitea, and according to the website it's not an explicitly supported upgrade path.

**Comment by strk** [comment]
@sbrl see also #1377 for setting users limits (and use the forum or IRC chat or file another issue to know about upgrade)
